### PR TITLE
Dont create parallel meta index inside the index

### PR DESCRIPTION
**Pull request by lava**
###  :notebook_with_decorative_cover: Description

See commit messages.

NOTE: After implementing this, I think it might be a good idea to rewrite this once more to give the `persist`-handler the following signature:

```
using active_partition_actor = caf::typed_actor<
    caf::replies_to<atom::persist>::with<atom::done, partition_synopsis>,
    [...]
```

Then the logic in the indexer would be to always forward expressions to all partitions that are currently known as `unpersisted`.

But let's discuss that first.

###  :memo: Checklist

- [x] All user-facing changes have changelog entries.
- [x] The changes are reflected on [docs.tenzir.com/vast](https://docs.tenzir.com/vast), if necessary.
- [x] The PR description contains instructions for the reviewer, if necessary.

### :dart: Review Instructions

Commit by commit

### CODE REVIEW

File: libvast/vast/system/index_actor.hpp [hunk 1]
```diff
@@ -38,11 +38,12 @@ using index_actor = caf::typed_actor<
   // Subscribes a FLUSH LISTENER to the INDEX.
   caf::reacts_to<atom::subscribe, atom::flush, wrapped_flush_listener>,
   // Evaluatates an expression.
-  caf::reacts_to<expression>,
+  caf::reacts_to<expression>, caf::reacts_to<expression, size_t>, // private
   // Queries PARTITION actors for a given query id.
   caf::reacts_to<uuid, uint32_t>,
   // Replaces the SYNOPSIS of the PARTITION witht he given partition id.
```
Comment (tobim) [inline libvast/vast/system/index_actor.hpp:44]:
```suggestion
  // Replaces the SYNOPSIS of the PARTITION with the given partition id.
```

File: libvast/vast/system/index_actor.hpp [hunk 2]
```diff
@@ -38,11 +38,12 @@ using index_actor = caf::typed_actor<
   // Subscribes a FLUSH LISTENER to the INDEX.
   caf::reacts_to<atom::subscribe, atom::flush, wrapped_flush_listener>,
   // Evaluatates an expression.
-  caf::reacts_to<expression>,
+  caf::reacts_to<expression>, caf::reacts_to<expression, size_t>, // private
```
Comment (tobim) [inline libvast/vast/system/index_actor.hpp:41]:
```suggestion
  caf::reacts_to<expression>, 
  // private
  caf::reacts_to<expression, size_t>,
```

File: libvast/vast/meta_index.hpp [hunk 3]
```diff
@@ -70,19 +70,13 @@ struct partition_synopsis {
...
-  /// @param slice The table slice to extract data from.
-  /// @param partition The partition ID that *slice* belongs to.
+  /// index. Only used in unit tests.
+  // TODO: Remove this and use `merge()` in unit tests instead.
   void add(const uuid& partition, const table_slice& slice);
 
-  /// Adds new synopses for a partition in bulk. Used when
-  /// re-building the meta index state at startup.
+  /// Adds new synopses for a partition in bulk.
   void merge(const uuid& partition, partition_synopsis&&);
```
Comment (tobim) [inline libvast/vast/meta_index.hpp:78]: This function should be called `add`.

File: libvast/src/system/index.cpp [hunk 4]
```diff
@@ -646,26 +656,30 @@ index(index_actor::stateful_pointer<index_state> self,
...
       // the unique owner of the pointer will be the message (which doesnt
       // need it anymore).
       // Semantically we want a unique_ptr here, but caf message types need
       // to be copy constructible.
       VAST_DEBUG(self, "replaces synopsis for partition", partition_id);
       if (!ps.unique()) {
-        VAST_WARNING(self, "ignores partition synopsis that is still in use");
-        // TODO: Should this return caf::skip?
-        return;
+        VAST_DEBUG(self, "skips partition synopsis that is still in use");
```
Comment (tobim) [inline libvast/src/system/index.cpp:673]: This is not supposed to happen, so it should still be a warning.

**Review by tobim — CHANGES_REQUESTED** [review]
Just a quick first pass, more to come later.

File: libvast/src/system/index.cpp [hunk 5]
```diff
@@ -50,6 +50,7 @@
 #include "vast/value_index.hpp"
 
 #include <caf/error.hpp>
+#include <caf/typed_response_promise.hpp>
```
Comment (dominiklohmann) [inline libvast/src/system/index.cpp:53]: This can be removed.

**Review by dominiklohmann — COMMENTED** [review]
Just looked at this out of curiosity. Note that we should not use platform-dependent types like `size_t` in actor interfaces. Instead, use `uint64_t` or, which would be even better, use a distinct type that wraps an unsigned integer instead.

**Comment by lava** [comment]
Closing this in favor of the superior design in https://github.com/tenzir/vast/pull/1296

**Pull request closed without merging**